File: 985-bag-of-tokens/bag-of-tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:

        tokens.sort()
        n = len(tokens)
        i,j=0,n-1
        score =0
        ans =0 
        while(i<=j):

            while(i<=j and tokens[i]<=power):
                power-=tokens[i]
                score+=1
                i+=1
                ans = max(ans,score)
            if score>=1:
                power+=tokens[j]
                score-=1
                j-=1
            else:
                break
                
        return ans

           

        

File: 985-bag-of-tokens/bag-of-tokens.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
          Arrays.sort(tokens);
        int n = tokens.length;
        int i = 0, j = n - 1;
        int score = 0;
        int ans = 0;

        while (i <= j) {
            while (i <= j && tokens[i] <= power) {
                power -= tokens[i];
                score++;
                i++;
                ans = Math.max(ans, score);
            }
            if (score >= 1) {
                power += tokens[j];
                score--;
                j--;
            } else {
                break;
            }
        }

        return ans;
    }
}